dhis2/dhis2-core
# feat: System update notification system

### Summary

This PR introduces automatic notification of new patch versions.
It works like this:
1. a Job cron trigger will trigger at CRON_DAILY_3AM_RANDOM_MIN_SEC set in SchedulerStart.java. We randomize the hour, minute and seconds to avoid all servers flooding the release document hosting server at the exact same time.
2. Fetch and parse this document: [https://releases.dhis2.org/v1/versions/stable.json]()
2. Compare parsed and sorted versions from JSON against the current version and find the most recent one.
3. If there is one or more newer patch versions, pick the latest.
4. Construct an internal message telling that there is a new version and send it to the system update recipient group, configured in the System Settings app. If there is no such group configured, a message will be sent til all users with the ALL authority.

NOTE: In step **4** we first check if the message has already been sent, to avoid duplicates. So the notification will only be sent ONE time for each new version. This check is performed in this line: **SystemUpdateService** - ``` List<MessageConversation> existingMessages = messageService.getMatchingExtId( version.getValue() );```
The extId is basically a unique ID for each version and is enforced globally for all users of the recipient group.

The system update notification user group can be set with the: **'/configuration/systemUpdateNotificationRecipients'** endpoint, similar to: **'/feedbackRecipients'**

This feature can be disabled with the dhis.conf configuration key: system.update_notifications_enabled

### Manually test:

There is a way to force trigger a check and sending of messages by using this API endpoint:
**http://localhost:8080/api/systemUpdates.json?forceVersion=2.36.1**

**forceVersion** is the version you want to fake in order to trick the check.

NB: Remember, it will only check for patch versions and the version must be a valid semantic version [https://semver.org/]().


JIRA:
[https://jira.dhis2.org/browse/DHIS2-9897](https://jira.dhis2.org/browse/DHIS2-9897)
